### PR TITLE
Have DiscussionXBlock and DiscussionCourseXBlock take care of loading assets they depend on

### PR DESCRIPTION
**Pull request by itsjeyd**
`DiscussionXBlock` currently relies on courseware to load JS and CSS assets that it depends on. This PR transfers responsibility for loading these assets to the XBlock itself.

`DiscussionCourseXBlock` was already taking care of loading *some* necessary assets. This PR makes sure that the block loads all assets it depends on.

cf. [YONK-382](https://openedx.atlassian.net/browse/YONK-382)

**Related PRs**

- https://github.com/edx/edx-platform/pull/13483

**Dependencies**

- https://github.com/edx-solutions/cs_comments_service/pull/18
- https://github.com/edx-solutions/jquery-xblock/pull/24

**Screenshots**

- Adding a post: add-a-post

- Expanded discussion: post-mathjax-expanded

- Responses: response-comment-mathjax

- Error message: post-with-error

**Setup instructions**

1. Get Studio, LMS, and Apros running on the parent branch (`ziafazal/fix-broken-features`), then checkout this branch. List of commits to use:

    * `edx-platform`: Latest commit on this branch.
    * `cs_comments_service`: https://github.com/edx-solutions/cs_comments_service/commit/10826f6c7b9a597c527bf9752f16567e5025d27e
    * Apros: `master` @ `b7e058b`

2. As `edxapp` user, remove static assets via `git clean -fx`.

3. As `edxapp` user, rebuild assets via `paver update_assets` *or* `paver devstack studio`, `paver devstack lms`.

4. Install changes from https://github.com/edx-solutions/jquery-xblock/pull/24: In Apros repo, replace contents of `static/js/vendor/jquery.xblock.js` with contents of [updated version](https://raw.githubusercontent.com/edx-solutions/jquery-xblock/2037d7411a063124bcc0e10615edfbd8db67c3b2/jquery.xblock.js).

5. As `apros` user, run `./manage.py collectstatic --noinput` and make sure `jquery.xblock.js` is mentioned in the output as one of the files that got updated.

**Test instructions**

1. Add `DiscussionXBlock` to a unit via the main "Discussion" component button.

2. Add `discussion-course` to advanced modules.

3. Create a new section called `DISCUSSION_TAB`.

4. Add a subsection and unit to the `DISCUSSION_TAB` section created in the previous step.

5. Select "Advanced" > "Discussion Course" to add an instance of `DiscussionCourseXBlock` to the unit.

6. Navigate to unit containing `DiscussionXBlock` in the LMS and smoke test discussion functionality.

    For example, make sure that you can perform the following actions as `staff` user:

    * Create a Post, Reply, Comment.
    * Edit a Post, Reply, Comment.
    * Test [MathJax](http://docs.mathjax.org/en/latest/start.html) functionality in Post, Reply, Comment (and Preview), with e.g.
      ```
There are two solutions to \(ax^2 + bx + c = 0\) and they are
$$x = {-b \pm \sqrt{b^2-4ac} \over 2a}.$$

When $a \ne 0$
      ```
    * Delete a Post, Reply, Comment.
    * Follow/unfollow a thread/post.
    * In the "Instructor" panel, add the `staff` user to the `Discussion Admin` or `Discussion Moderator` group. Then you should be able to: 
        * Open/close a thread/post.
        * Pin/unpin a thread/post.
    * Try to create a post without a title. This should cause an error message to be displayed above the controls you used to create the post.

6. Login as a regular student.  You should be able to do the above actions, plus:

    * Report/unreport a post created by the `staff` user.
    * Vote/unvote a thread/post created by the `staff` user.

7. Repeat previous steps for unit containing `DiscussionCourseXBlock`.

8. In Apros, navigate to unit containing `DiscussionXBlock` and repeat previous steps. Performing some of the actions will be a little fiddly because CSS for `DiscussionXBlock` does not match Apros CSS yet (cf. "Author notes" below).

  

9. In Apros, navigate to (equivalent of) "Discussion" tab and repeat previous steps. The CSS for `DiscussionCourseXBlock` does not match Apros CSS yet (cf. "Author notes" below), so some UI elements will look a little weird.

  

10. Enable the "Teams" feature:

    * In Studio, replace the contents of "Advanced Settings" > "Teams Configuration" with:

      ```json
{
    "topics": [
        {
            "name": "Sustainability in Corporations",
            "description": "Description for Sustainability in Corporations",
            "id": "Sustain_Corporations"
        },
        {
            "name": "Water Conservation Projects",
            "description": "Description for Water Conservation",
            "id": "Water_Conservation"
        },
        {
            "name": "Sustainability Standards and Reporting",
            "description": "Description for Sustainability Standards",
            "id": "Standards_Reporting"
        }
    ],
    "max_team_size": 5
}
      ```
    * In the LMS, navigate to the "Teams" tab, select one of the topics and add a team to it via "create a new team in this topic".
    * Sanity-check discussion CSS by adding a post, response, and comment.

**Author notes**

- CSS for both `DiscussionXBlock` and `DiscussionCourseXBlock` does not match Apros CSS yet. We have a separate ticket for addressing this: [YONK-383](https://openedx.atlassian.net/browse/YONK-383).

**Reviewers**

- [x] OpenCraft: @pomegranited 
- [ ] Solutions: @ziafazal

### CODE REVIEW

File: common/static/common/js/discussion/discussion_router.js [hunk 1]
```diff
@@ -127,17 +166,14 @@
...
 
             DiscussionRouter.prototype.navigateToThread = function(thread_id) {
-                var thread;
+                var thread, targetThreadRoute;
                 thread = this.discussion.get(thread_id);
-                return this.navigate("" + (thread.get("commentable_id")) + "/threads/" + thread_id, {
-                    trigger: true
-                });
+                targetThreadRoute = getSingleThreadRoute(thread.get("commentable_id"), thread_id);
+                this.navigate(targetThreadRoute, { trigger: true});
```
Comment (pomegranited) [inline common/static/common/js/discussion/discussion_router.js]: nit: remove space between `{` and `trigger`.

File: common/static/common/js/discussion/discussion_router.js [hunk 2]
```diff
@@ -127,17 +166,14 @@
...
             };
 
             DiscussionRouter.prototype.navigateToAllThreads = function() {
-                return this.navigate("", {
-                    trigger: true
-                });
+                this.navigate(allThreadsRoute, { trigger: true });
```
Comment (pomegranited) [inline common/static/common/js/discussion/discussion_router.js]: nit: remove spaces between `{` and `trigger`, and between `true` and `}`.

File: openedx/core/lib/xblock_builtin/xblock_discussion/static/js/mathjax_include.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+var vendorScript;
+if (typeof MathJax === 'undefined') {
+    vendorScript = document.createElement('script');
+    vendorScript.onload = function() {
+        'use strict';
+        var MathJax, setMathJaxDisplayDivSettings;
```
Comment (pomegranited) [inline openedx/core/lib/xblock_builtin/xblock_discussion/static/js/mathjax_include.js]: This `var MathJax` declaration is masking the `window.MathJax` that was defined by the loaded script, and so `MathJax` ends up `undefined`.  Please change this line to:

`var setMathJaxDisplayDivSettings;`

Or alternately, set `var MathJax = window.MathJax`;

**Comment by pomegranited** [comment]
:+1: Once these issues are addressed:

* Nit: In the Test Instructions, I had to add `discussion-course` (not `discussion-forum`) to the Advanced Modules to enable `DiscussionCourseXBlock`.
* Two code nits ([1](https://github.com/edx-solutions/edx-platform/pull/744/files#r80403929), [2](https://github.com/edx-solutions/edx-platform/pull/744/files#r80403992))
* [One JS fix](https://github.com/edx-solutions/edx-platform/pull/744/files#r80409085).

- [x] I tested this using the instructions in the PR.  It worked as described.
- [x] I read through the code - Thank you for adding that separate branch!  Made reviewing your changes much easier (even if I did keep forgetting where to leave my comments! :smile:)
- [x] I checked for accessibility issues by ensuring I could reach all functionality via my keyboard.
- [x] Includes documentation: docstrings are sufficient, no client-facing documentation updates required.

File: openedx/core/lib/xblock_builtin/xblock_discussion/static/js/mathjax_include.js (hunk 3 above)
Comment (itsjeyd) [inline openedx/core/lib/xblock_builtin/xblock_discussion/static/js/mathjax_include.js]: @pomegranited Doh... nice catch. That's what happens when it's late and you're just trying to make the JS linter happy ;)

**Comment by itsjeyd** [comment]
@pomegranited Thanks for the review! I addressed your comments, updated the PR description, and integrated the commit (https://github.com/edx-solutions/edx-platform/pull/743/commits/2a22dba2bd19df1b5517713441463b777d53fcd6) that you added to https://github.com/edx-solutions/edx-platform/pull/743.

@ziafazal This is ready for you to have a look. It builds on https://github.com/edx-solutions/edx-platform/pull/743. To only see the changes introduced here, see this commit: https://github.com/open-craft/edx-platform/commit/117350e2aad0aacf7e6218b65c0bdab2faf33d38. Thanks!

File: openedx/core/lib/xblock_builtin/xblock_discussion/static/js/URI_shim.js [hunk 4]
```diff
@@ -1,8 +1,5 @@
 if (typeof URI === 'undefined') {
-    console.log('URI is not defined. Loading it now ...');
```
Comment (e-kolpakov) [inline openedx/core/lib/xblock_builtin/xblock_discussion/static/js/URI_shim.js]: @itsjeyd nit: actually it would be great to still have those log messages - it could help debugging when something goes wrong.

**Comment by itsjeyd** [comment]
@ziafazal FYI, I rebased this branch, so the diff is clean now.

**Comment by ziafazal** [comment]
@itsjeyd I started testing in-line discussion block in apros and got list of JS errors and i think because of these JS error `Add a Post` did not work. PFA output of console.
[apros.mcka.txt](https://github.com/edx-solutions/edx-platform/files/495021/apros.mcka.txt)
Am I missing something?

**Comment by itsjeyd** [comment]
@ziafazal I can't reproduce this with the setup from the PR description. But I remember getting a similar error ("Uncaught Error: Mismatched anonymous define() module" in the LMS when I was working on the implementation a few days ago. If I remember correctly, it was related to a dependency getting loaded twice. You could try reverting this commit for debugging purposes: https://github.com/edx-solutions/edx-platform/pull/744/commits/0a2af9c33db5d3c2e919e51b0399d03a1773c885. That might give some indication as to what's going on. When loading a unit with an instance of `DiscussionXBlock` in the LMS, you should see this in the browser console:

```
URI is defined. Proceeding.
RequireJS is defined. Proceeding.
edx namespace is defined. Proceeding.
```

In Apros, you should see the opposite:

```
URI is not defined. Loading it now ...
RequireJS is not defined. Loading it now ...
edx namespace is not defined. Loading it now ...
```

Could you try that and report back?

**Comment by itsjeyd** [comment]
@ziafazal FYI: I updated this PR with a fix for CSS regressions affecting "Teams", cf. https://github.com/edx/edx-platform/pull/13483#issuecomment-249576704.

CC @pomegranited

File: openedx/core/lib/xblock_builtin/xblock_discussion/static/js/discussion_classes.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+            }
+        );
+    });
+};
+
+if (typeof RequireJS === 'undefined') {
+    var vendorScript = document.createElement("script");
+    vendorScript.onload = function() {
+        'use strict';
+        defineDiscussionClasses(define);
```
Comment (ziafazal) [inline openedx/core/lib/xblock_builtin/xblock_discussion/static/js/discussion_classes.js]: @itsjeyd  I'm still seeing those JS errors. after commenting this line i got rid of this error
```
require.js:165 Uncaught Error: Mismatched anonymous define() module
```
It looks like

Comment (itsjeyd) [inline openedx/core/lib/xblock_builtin/xblock_discussion/static/js/discussion_classes.js]: @ziafazal This is a bit difficult for me to debug, since I can't reproduce the problem. Could you please tell me what your exact setup is (in terms of commits and anything else that you think would be useful to know)?

BTW,  I didn't mean to suggest in my previous [comment](https://github.com/edx-solutions/edx-platform/pull/744#issuecomment-249863368) that reverting https://github.com/edx-solutions/edx-platform/pull/744/commits/0a2af9c33db5d3c2e919e51b0399d03a1773c885 would fix the problem on your end (if that's the impression you got). What I meant was that reverting the commit might help with debugging the issue. When you load a unit with an instance of `DiscussionCourseXBlock` in Apros, do you get the following output (or do some of these messages appear multiple times)?

```
URI is not defined. Loading it now ...
RequireJS is not defined. Loading it now ...
edx namespace is not defined. Loading it now ...
```

**Comment by ziafazal** [comment]
@itsjeyd overall it LGTM except one "Expand Discussion" feature is not working for inline discussions both in LMS and Apros. See screenshot it gives 404 in console.
screen shot 2016-09-29 at 11 12 07 am

**Comment by itsjeyd** [comment]
@ziafazal Thanks for the review! We ran into the same problem with "Expand discussion" prior to #743 (see [YONK-401](https://openedx.atlassian.net/browse/YONK-401), but that PR fixed the issue. I haven't seen it since. Could you please list the steps that you performed before clicking "Expand discussion"? Also, just to confirm, do you get this error for every post, or just for a specific one?

**Comment by ziafazal** [comment]
@itsjeyd it looks this PR's branch is rebased with branch of PR #743 why we still have this issue?.

1. I added an inline discussion to course and that course has only two sections one has course discussion and other has inline discussion.

2. Added a new post with some tex and mathjax expression given given in instructions of this PR.

3. Added text response to the post.

4. Reloaded the page and clicking on "Expand Discussion" shows modal dialog with 404 error in console.

Yeah I'm getting error on  expanding every post.

**Comment by itsjeyd** [comment]
Thanks @ziafazal for the additional info. I did some more testing, but I can't reproduce the problem. Clicking "Expand discussion" works fine in both LMS and Apros; I don't get the modal from your screenshot, and there are no errors in the console.

Following the steps you listed in your [comment](https://github.com/edx-solutions/edx-platform/pull/744#issuecomment-250395333), the workflow I tested went like this:

* Load unit with instance of `DiscussionXBlock`.
* Click "Add a Post".
* Enter text for testing MathJax from PR description.
* Click "Add Post".
* Click "Expand discussion" to make the form for adding a response visible. *Worked without issue.*
* Enter response in "Add a response:" field.
* Click "Submit".
* Reload the page.
* Click "Show discussion".
* Click "Expand discussion" on the newly added post. *Worked without issue.*

**Comment by ziafazal** [comment]
@itsjeyd if you are unable to reproduce then it appears issue might be related to my environment. I'm good to merge this PR.

**Comment by itsjeyd** [comment]
Thanks @ziafazal! A quick heads-up that we'll be following up with some additional changes from https://github.com/edx/edx-platform/pull/13483 soon, to minimize the differences between the changes introduced in the two PRs.

**Comment by pomegranited** [comment]
@ziafazal I've cherry-picked @itsjeyd 's changes from [the upstream PR](https://github.com/edx/edx-platform/pull/13483), and added a couple more changes to remove some code duplication.

Could you have another look, and let us know if the changes are ok to merge?

**Comment by ziafazal** [comment]
@pomegranited  @itsjeyd LGTM 👍

**Comment by pomegranited** [comment]
Thanks, @ziafazal!  Squashing and merging now.
